File: leetcode/partition.py
class Solution:
    # https://support.leetcode.com/hc/en-us/articles/360011834174-I-encountered-Wrong-Answer-Runtime-Error-for-a-specific-test-case-When-I-test-my-code-using-this-test-case-it-produced-the-correct-output-Why-
    def __init__(self):
        self.output = []

    def is_palindrome(self, s: str):

        left = 0
        right = len(s) -1

        while left < right:
            if s[left] != s[right]:
                return False
            left += 1
            right -= 1

        return True

    def get_partition(self, s : str, res : List[str]):

        if not s:
            self.output.append(res)

        for i in range(0, len(s)):
            prefix = s[0:i + 1]
            suffix = s[i + 1:]

            if self.is_palindrome(prefix):
                # make a deep copy
                temp_res = res[:]
                temp_res.append(prefix)

                self.get_partition(suffix, temp_res)


    def partition(self, s: str) -> List[List[str]]:
        self.get_partition(s, [])
        return self.output

File: leetcode/Partition.java
class Solution {
    
    public boolean isPalindrome(String s){
        int n = s.length();
        
        for (int i = 0 ; i < (n / 2); i++){
            if (s.charAt(i) != s.charAt(n - i -1)){
                return false;
            }
        }
        return true;
    }
    
    List<List<String>> output = new ArrayList<>();
    
    public void getPartition(String s, List<String> res){
        
        if (s.isEmpty()){
           output.add(res);
           return;
        }
        
        for (int i = 0; i < s.length(); i++){
            String prefix = s.substring(0, i + 1);
            String suffix = s.substring(i + 1);
            
            if (isPalindrome(prefix)){
                List<String> tempRes = new ArrayList<>(res);
                tempRes.add(prefix);
                getPartition(suffix, tempRes);
            }
        }
        
    }
    
    public List<List<String>> partition(String s) {        
        
        List<String> res = new ArrayList<>();
        
        getPartition(s, res);
        
        return output;
    }
}